### PR TITLE
Update Creating a Module guide

### LINKED ISSUE #5103: Expand 'How to create Loris module' wiki
How to make a LORIS module page should be expanded to cover the following parts of adding a new module to Loris:

JavaScript: use getJSDependencies() - see example in PR#1632 )
Add Permissions
Recommend View and Edit.
Consider site interactions: Should users only be able to see content from other sites if they have Access All Profiles permission?
How to modify hasAccess().
Add to the LorisMenu and LorisMenuPermissions table.
Adding Configuration settings: How to add to ConfigSetting table and then how to populate the value in the Config table (Patch and schema change).
Adding dropdown help text
How to do the SQL for this (patch and schema change)
Caveats like: "Note interactions in with key configuration variables like $useProject, when coding your queries"

### PR DESCRIPTION
**Pull request by johnsaigle**
Creates a guide for creating a modern LORIS module and adds it to Read The Docs.

Also moves the Help Text Style guide to the Developers section of the wiki.

Fixes #5103

### CODE REVIEW

**Comment by johnsaigle** [comment]
Note: this needs to add more points request by #5103

File: docs/wiki/99 - Developers/Creating a Module.md [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+It can also optionally include widgets to be displayed on the dashboard.
+
+* A PHP file with the same name as the module: `php/my_new_module.class.inc`.
+
+This class contains the main functionality of module, defining complex user permissions,
+handling of requests, and definining JavaScript and CSS dependencies. 
+
+[See this file for an example](../../../modules/dicom_archive/php/dicom_archive.class.inc).
+
+Additiaonl PHP files will depend heavily on the kind of module you are creating. The most basic types
```
Comment (PapillonMcGill) [inline docs/wiki/99 - Developers/Creating a Module.md]:
```suggestion
Additional PHP files will depend heavily on the kind of module you are creating. The most basic types
```

File: docs/wiki/99 - Developers/Creating a Module.md [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+of LORIS modules are "Menu Filters" and "Forms".
+
+#### Menu Filters
+
+This is a type of module that displays a table of data with options to filter the values of this table.
+
+The correct way to create a menu filter is to include additional classes that represent the idea of a
+row in the tabel displayed to a user. This differs from a row in the database itself; for example, a row in the
```
Comment (PapillonMcGill) [inline docs/wiki/99 - Developers/Creating a Module.md]:
```suggestion
row in the table displayed to a user. This differs from a row in the database itself; for example, a row in the 
```

File: docs/wiki/99 - Developers/Creating a Module.md [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+
+A Loris module consists of a group of pages that perform a specific function or assist with a specific task. In Loris, all active modules are shown as submenu items in the main menu on top of the page. Examples of modules include `Document Repository`, `Candidate List (Access Profiles)`, `Imaging Browser` ...
+
+
+## Directory structure
+
+1. Create a new folder corresponding to your module name under `modules/`. 
+This name should be short and descriptive. 
+The name must use only lowercase alphanumeric characters and underscores, e.g. `document_repository`.
+For the remainder of this document we will assume your new module is named `my_first_module`.
```
Comment (PapillonMcGill) [inline docs/wiki/99 - Developers/Creating a Module.md]:
```suggestion
For the remainder of this document we will assume your new module is named `my_new_module`.
```
To reflect usage of `my_new_module` in the rest of this file

File: docs/wiki/99 - Developers/Creating a Module.md [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+You should create a class `php/mynewmodulerow.class.inc` to represent a Menu Filter row. This will
+provide basic structure for the row and include permission flags that can restrict data to users
+affiliated with a given Site or Project.
+
+
+[See this file for an example](../../../modules/dicom_archive/php/dicomarchiverow.class.inc).
+
+##### Row Provisioner.
+This file should be accompanied by a class `php/mynewmodulerowprovisioner.class.inc`. This file should
+conntain the SQL statement used to query the database for data that will be consolidated into a
```
Comment (PapillonMcGill) [inline docs/wiki/99 - Developers/Creating a Module.md]:
```suggestion
contain the SQL statement used to query the database for data that will be consolidated into a
```

File: docs/wiki/99 - Developers/Creating a Module.md [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+### SQL Files
+
+New modules must be inserted into the database. You must create new permissions for the module and insert
+an entry for your new module within the `modules` table in LORIS.
+
+#### Module permissions
+
+You must define new permissions for the module. These are used to limit users to specific actions with the module.
+
+Permissions are defined in the `permissions` table in LORIS and should be separateed into view and edit permissions in most cases.
```
Comment (PapillonMcGill) [inline docs/wiki/99 - Developers/Creating a Module.md]:
```suggestion
Permissions are defined in the `permissions` table in LORIS and should be separated into view and edit permissions in most cases.
```

File: docs/wiki/99 - Developers/Creating a Module.md [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+
+The remaining PHP files will depend heavily on the kind of module you are creating. The most basic types
+of LORIS modules are "Menu Filters" and "Forms".
+
+#### Menu Filters
+
+This is a type of module that displays a table of data with options to filter the values of this table.
+
+The correct way to create a menu filter is to include additional classes that represent the idea of a
+row in the tabel displayed to a user. This differs from a row in the database itself; for example, a row in the
```
Comment (johnsaigle) [inline docs/wiki/99 - Developers/Creating a Module.md]:
```suggestion
row in the table displayed to a user. This differs from a row in the database itself; for example, a row in the 
```

File: docs/wiki/99 - Developers/Creating a Module.md [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+Modern LORIS modules use ReactJS to present data. The content of these files will vary significantly
+depending on what functionality you want to provide. The best way to learn about creating JavaScript
+for a new module is by studying the code of an existing module that is similar to what you are
+trying to design.
+
+Be sure to familiarize yourself with our [ReactJS guidelines](./ReactJS Guidelines.md).
+
+### SQL Files
+
+New modules must be inserted into the database. You must create new permissions for the module and insert
```
Comment (PapillonMcGill) [inline docs/wiki/99 - Developers/Creating a Module.md]: a little bit redundant with what follow in `Module permissions`

File: docs/wiki/99 - Developers/Creating a Module.md [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+Permissions are defined in the `permissions` table in LORIS and should be separated into view and edit permissions in most cases.
+
+In our example, you would create two new permissions: `my_new_module_view` and `my_new_module_edit`. Your PHP code within
+`php/my_new_module.class.inc` will define what these permissions allow a user to do using the `hasAccess()` function. 
+
+To add new permissions, append them to the list of existing permissions found in `SQL/0000-00-01-Permission.sql`. (This file is found in your LORIS root directory, not within the module subdirectory.)
+
+#### Modules table
+
+All modules must be listed in the file `SQL/0000-00-02-Modules.sql`.
```
Comment (PapillonMcGill) [inline docs/wiki/99 - Developers/Creating a Module.md]: Should add a link to the Database schema modeling standard. https://github.com/aces/Loris/blob/master/docs/SQLModelingStandard.md

File: docs/wiki/99 - Developers/Creating a Module.md [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+#### Modules table
+
+All modules must be listed in the file `SQL/0000-00-02-Modules.sql`. 
+
+Include a new entry in this file by adding to the list of existing modules.
+
+#### Creating a patch
+
+The files referenced above will be run automatically for new LORIS installations. In addition to these files,
+you msut create a "patch" file that existing LORIS instances can run in order to use your new module.
```
Comment (PapillonMcGill) [inline docs/wiki/99 - Developers/Creating a Module.md]:
```suggestion
you must create a "patch" file that existing LORIS instances can run in order to use your new module.
```

File: docs/wiki/99 - Developers/Creating a Module.md (hunk 7 above)
Comment (johnsaigle) [reply docs/wiki/99 - Developers/Creating a Module.md]: Module Permissions is a subheading of this section. It's intended to expand upon the general idea of this sentence.

File: docs/wiki/99 - Developers/Creating a Module.md [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+to `SQL/0000-00-01-Permission.sql` and `SQL/0000-00-02-Modules.sql`.
+
+### Accessing the module
+
+The new module must be included in the modules table 
+Once you created the modules as described above, go to `$LORIS$/$ModuleName$` to see your newly created page
+
+### Testing
+
+Your module must contain a file `test/TestPlan.md` that enumerates a sequence of action that a developer can
```
Comment (PapillonMcGill) [inline docs/wiki/99 - Developers/Creating a Module.md]:
```suggestion
Your module must contain a file `test/TestPlan.md` that enumerates a sequence of actions that a developer can
```

File: docs/wiki/99 - Developers/Creating a Module.md [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+follow to test all functionality of your module.
```
Comment (PapillonMcGill) [inline docs/wiki/99 - Developers/Creating a Module.md]:
```suggestion
follow to test all functionalities of your module.
```

File: docs/wiki/99 - Developers/Creating a Module.md [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+
+A new module must contain a `README.md` file that briefly describes the functionality of the module.
+Examples can be found in any existing module.
+
+It must also contain help text that will be displayed to users. In contrast to the README, this should
+be written in a friendly and non-technical tone. Please see the [Help Text Style Guide](./Help Style Guide.md).
+
+### Issuing a pull request
+
+A pull request containing a new module `my_new_module` will lmust contain the following
```
Comment (PapillonMcGill) [inline docs/wiki/99 - Developers/Creating a Module.md]:
```suggestion
A pull request containing a new module `my_new_module` must contain the following 
```

File: docs/wiki/99 - Developers/Creating a Module.md [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+* `modules/my_new_module/help/my_new_module.md` -- Front-end help text for users
+* `modules/my_new_module/jsx/...` -- File(s) containing ReactJS code
+* `modules/my_new_module/js/` -- Empty, will contain compiled JS locally
+* `modules/my_new_module/php/module.class.inc` -- Defining module metadata, widgets, and basic permissions
+* `modules/my_new_module/php/my_new_module.class.inc` -- Containing core back-end functionality
+* `modules/my_new_module/php/...` -- Additional PHP files as needed
+* `modules/my_new_module/test/TestPlan.md` -- An exhaustive list of steps to test functionality
+* `modules/my_new_module/README.md` -- Technical summary of module
+* `modules/my_new_module/.gitignore` -- Defines local files to be ignored by git; usually contains `js/*`
+* e.g. `SQL/New_patches/$date_Add-New-Module.sql` -- SQL statements to modify an existing LORIS to be compatible with the new module.
```
Comment (PapillonMcGill) [inline docs/wiki/99 - Developers/Creating a Module.md]:
```suggestion
* e.g. `SQL/New_patches/$date_Add-New-Module.sql` -- SQL statements to modify an existing LORIS instance to be compatible with the new module.
```

File: docs/wiki/99 - Developers/Creating a Module.md [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+depending on what functionality you want to provide. The best way to learn about creating JavaScript
+for a new module is by studying the code of an existing module that is similar to what you are
+trying to design.
+
+Be sure to familiarize yourself with our [ReactJS guidelines](./ReactJS Guidelines.md).
+
+### SQL Files
+
+New modules must be inserted into the database. You must create new permissions for the module and insert
+an entry for your new module within the `modules` table in LORIS.
```
Comment (johnsaigle) [inline docs/wiki/99 - Developers/Creating a Module.md]:
```suggestion
an entry for your new module within the `modules` table in LORIS.

Please make sure that all SQL changes follow our [SQL code standards](https://github.com/aces/Loris/blob/master/docs/SQLModelingStandard.md).
```

File: docs/wiki/99 - Developers/Creating a Module.md (hunk 8 above)
Comment (johnsaigle) [reply docs/wiki/99 - Developers/Creating a Module.md]: Added on line 98

File: docs/wiki/99 - Developers/Creating a Module.md (hunk 7 above)
Comment (PapillonMcGill) [reply docs/wiki/99 - Developers/Creating a Module.md]: This section (specially the 2nd phrase) contain details that should be in sub-section.

I would change this section to something like:
```
A pull request to create a new module must contain SQL files to let LORIS detect the module and create appropriate permissions.
```
Then in the `Modules table` below, introduce the `modules` table.

File: docs/wiki/99 - Developers/Creating a Module.md [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+You should create a class `php/mynewmodulerow.class.inc` to represent a Menu Filter row. This will
+provide basic structure for the row and include permission flags that can restrict data to users
+affiliated with a given Site or Project.
+
+
+[See this file for an example](../../../modules/dicom_archive/php/dicomarchiverow.class.inc).
+
+##### Row Provisioner.
+This file should be accompanied by a class `php/mynewmodulerowprovisioner.class.inc`. This file should
+contain the SQL statement used to query the database for data that will be consolidated into a
```
Comment (driusan) [inline docs/wiki/99 - Developers/Creating a Module.md]: Provisioners don't need to be SQL statements, they just often are.

File: docs/wiki/99 - Developers/Creating a Module.md [hunk 16]
```diff
@@ -0,0 +1,171 @@
...
+Permissions are defined in the `permissions` table in LORIS and should be separated into view and edit permissions in most cases.
+
+In our example, you would create two new permissions: `my_new_module_view` and `my_new_module_edit`. Your PHP code within
+`php/my_new_module.class.inc` will define what these permissions allow a user to do using the `hasAccess()` function. 
+
+To add new permissions, append them to the list of existing permissions found in `SQL/0000-00-01-Permission.sql`. (This file is found in your LORIS root directory, not within the module subdirectory.)
+
+#### Modules table
+
+All modules must be listed in the file `SQL/0000-00-02-Modules.sql`.
```
Comment (driusan) [inline docs/wiki/99 - Developers/Creating a Module.md]: I think it's better to point them to the script that manages the table. They shouldn't have to know anything about the table schema to add their module. It only needs to be added to the table in the default schema in order to show up in the menu on a default install.

File: docs/wiki/99 - Developers/Creating a Module.md [hunk 17]
```diff
@@ -0,0 +1,171 @@
...
+
+### Issuing a pull request
+
+A pull request containing a new module `my_new_module` must contain the following 
+
+#### Required new files:
+
+* `modules/my_new_module/help/my_new_module.md` -- Front-end help text for users
+* `modules/my_new_module/jsx/...` -- File(s) containing ReactJS code
+* `modules/my_new_module/js/` -- Empty, will contain compiled JS locally
```
Comment (driusan) [inline docs/wiki/99 - Developers/Creating a Module.md]: This isn't possible, git can't track empty directories..

File: docs/wiki/99 - Developers/Creating a Module.md [hunk 18]
```diff
@@ -7,57 +7,35 @@ A Loris module consists of a group of pages that perform a specific function or
 
 ## Directory structure
 
-1. Create a new folder corresponding to your module name under `modules/`. 
+Create a new folder corresponding to your module name under `modules/`. 
 This name should be short and descriptive. 
-The name must use only lowercase alphanumeric characters and underscores, e.g. `document_repository`.
+The name must should use only lowercase alphanumeric characters and underscores, e.g. `document_repository`.
```
Comment (PapillonMcGill) [inline docs/wiki/99 - Developers/Creating a Module.md]:
```suggestion
The name should use only lowercase alphanumeric characters and underscores, e.g. `document_repository`.
```

File: docs/wiki/99 - Developers/Creating a Module.md [hunk 19]
```diff
@@ -0,0 +1,147 @@
...
+
+
+[See this file for an example](../../../modules/dicom_archive/php/dicomarchiverow.class.inc).
+
+##### Row Provisioner.
+This file should be accompanied by a class `php/mynewmodulerowprovisioner.class.inc`. Provisioners usually
+contain the SQL statement(s) used to query the database for data that will be consolidated into a
+Menu Filter row.
+
+[See this file for an example](../../../modules/dicom_archive/php/dicomarchiverowprovisioner.class.inc).
```
Comment (johnsaigle) [inline docs/wiki/99 - Developers/Creating a Module.md:56]: Note to self: I don't think it's possible to traverse up from this directory into the modules/ directory. 

Absolute links will probably be necessary.

Comment (laemtl) [reply docs/wiki/99 - Developers/Creating a Module.md]: It is now.

**Comment by laemtl** [comment]
@driusan this is ready for review

File: docs/wiki/99_Developers/Creating_Module.md [hunk 20]
```diff
@@ -0,0 +1,145 @@
...
+### PHP Files
+
+The following files are required for the module to load in LORIS:
+
+* A PHP file called "module": `php/module.class.inc`.
+* A PHP file with the same name as the module: `php/my_new_module.class.inc`.
+
+The namespace must be set to `LORIS\$modulename` for anything defined in the PHP directory. Filenames need to be lowercase in order to be autoloaded. Class names can be any case.
+
+The remaining PHP files will depend heavily on the kind of module you are creating.
```
Comment (christinerogers) [inline docs/wiki/99_Developers/Creating_Module.md:24]:
```suggestion
The remaining PHP files will depend heavily on the kind of module you are creating. See the Imaging Browser or DICOM Archive for examples of modules with subpages. 
```

**Comment by johnsaigle** [comment]
@christinerogers I think this document should direct developers to create new modules in the best possible way. For that reason, I don't want to add a line about subpages because I think subpages aren't indicative of good design. In fact I think the recent permission issues with subpages are actually bad code smells: we shouldn't need to re-implement permissions this much. 

Instead of documenting subpages here and thereby implying that they are an intentional part of our design process going forward, we should probably refactor those modules. In general, our documentation shouldn't be apologies for our bugs especially in the context of explaining our ideal design principles.

A better approach would be to create a different document explaining project permissions in detail (either on the new wiki or on Google Drive). The complexities with subpages could be included there as a footnote.

I think it would more sense for someone who has written a module or dealt with many extensively to takeover the PR.

**Comment by christinerogers** [comment]
@johnsaigle sounds fair.  pointing to a separate section on permissions is probably wise.  The practical consideration though is - it is something people will need to know to do for the next few cycles as we work on the redesign/refactor.   If clarifying the rewording so that it doesn't sound like intentional design (which wasn't actually what i mean, just a caveat) is an option -- i'm all for that too.

File: docs/wiki/99_Developers/Creating_Module.md [hunk 21]
```diff
@@ -0,0 +1,145 @@
...
+menu filter module may represent a joining of columns across many database tables.
+
+If you are creating a Menu Filter page, your class `php/my_new_module.class.inc` should extend
+the class `\DataFrameworkMenu`.
+
+[See this file for an example](../../../modules/dicom_archive/php/dicom_archive.class.inc).
+
+##### Data Instance
+You should create a class `php/mynewmodulerow.class.inc` to represent a Menu Filter row. This will
+provide basic structure for the row and include permission flags that can restrict data to users
```
Comment (driusan) [inline docs/wiki/99_Developers/Creating_Module.md:44]: I don't know what "permission flags" are

File: docs/wiki/99_Developers/Creating_Module.md [hunk 22]
```diff
@@ -0,0 +1,145 @@
...
+
+We strongly encourage that new modules include integration tests. These must be written in the file
+`test/my_new_moduleTest.php`.
+
+### Documentation
+
+A new module must contain a `README.md` file that briefly describes the functionality of the module.
+Examples can be found in any existing module.
+
+It must also contain help text that will be displayed to users. In contrast to the README, this should
```
Comment (driusan) [inline docs/wiki/99_Developers/Creating_Module.md:120]: This should say where the help text document(s) go.

File: docs/wiki/99_Developers/Creating_Module.md [hunk 23]
```diff
@@ -0,0 +1,145 @@
...
+You should create a class `php/mynewmodulerow.class.inc` to represent a Menu Filter row. This will
+provide basic structure for the row and include permission flags that can restrict data to users
+affiliated with a given Site or Project.
+
+[See this file for an example](../../../modules/dicom_archive/php/dicomarchiverow.class.inc).
+
+##### Row Provisioner.
+This file should be accompanied by a class `php/mynewmodulerowprovisioner.class.inc`. Provisioners usually
+contain the SQL statement(s) used to query the database for data that will be consolidated into a
+Menu Filter row.
```
Comment (driusan) [inline docs/wiki/99_Developers/Creating_Module.md:52]: This should clarify that data can be provisioned by any source, and doesn't need to come from an SQL database.

Comment (johnsaigle) [reply docs/wiki/99_Developers/Creating_Module.md:52]: I don't really understand what this means -- doesn't all data in LORIS come from MySQL? Could you maybe use the suggest feature to make the changes you'd like to see?

File: docs/wiki/99_Developers/Creating_Module.md (hunk 22 above)
Comment (johnsaigle) [inline docs/wiki/99_Developers/Creating_Module.md:120]:
```suggestion
The new module must also a markdown file under `help/` that will be displayed to users. In contrast to the README, this should
```

File: docs/wiki/99_Developers/Creating_Module.md (hunk 21 above)
Comment (johnsaigle) [inline docs/wiki/99_Developers/Creating_Module.md:44]:
```suggestion
provide basic structure for the row and will likely restrict data access to users
```

**Comment by johnsaigle** [comment]
> and I’ll pick up #6048 after you leave because it’s just the wording about provisioners and data instances that need to be made more accurate and not assume that data is coming from SQL but that’s more than a slight tweaking of the wording.

(from Slack) Marking this as Needs Maintainer but also assigning to Dave

**Comment by driusan** [comment]
Replaced by #6833

**Pull request closed without merging**